packages/flutter/test/widgets/slivers_keepalive_test.dart

```diff
@@ -0,0 +1,564 @@
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Directionality(
+      textDirection: TextDirection.ltr,
+      child: Center(
+        child: Container(
+          height: 100,
+          child: CustomScrollView(
+            cacheExtent: 0,
+            slivers: <Widget>[
+              SliverFillViewport(
+                delegate: SliverChildBuilderDelegate(
+                  (BuildContext context, int index) {
+                    return children[index];
+                  },
+                  childCount: children.length,
+                  findChildIndex: (Key key) {
+                    return _mapKeyToIndex[key] == null ? -1 : _mapKeyToIndex[key];
+                  }
+                ),
+              )
+            ],
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+class SwitchingChildListTest extends StatefulWidget {
+  SwitchingChildListTest(this.children, {
+    this.viewportFraction = 1.0,
+    Key key}) : super(key: key);
+
+  final List<Widget> children;
+  final double viewportFraction;
+
+  @override
+  _SwitchingChildListTest createState() {
+    return _SwitchingChildListTest();
+  }
+}
+
+class _SwitchingChildListTest extends State<SwitchingChildListTest> {
+  List<Widget> children;
+
+  void switchChildrenOrder(int i, int j) {
+    setState(() {
+      children = List<Widget>.from(children);
+      final Widget w = children[i];
+      children[i] = children[j];
+      children[j] = w;
+    });
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    children = widget.children;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Directionality(
+      textDirection: TextDirection.ltr,
+      child: Center(
+        child: Container(
+          height: 100,
+          child: CustomScrollView(
+            cacheExtent: 0,
+            slivers: <Widget>[
+              SliverFillViewport(
+                viewportFraction: widget.viewportFraction,
+                delegate: SliverChildListDelegate(children),
+              )
+            ],
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+
+class WidgetTest1 extends StatefulWidget {
+  WidgetTest1(this.text, {
+    this.optionsRef = const <String, bool> {
+      'keepAlive': false,
+      'shouldDispose': true
+    },
+    Key key}) : super(key: key);
+
+  final String text;
+  final Map<String, bool> optionsRef;
+
+  @override
+  String toString({DiagnosticLevel minLevel = DiagnosticLevel.debug}) {
+    return 'WidgetTest1 $text';
+  }
+  @override
+  _WidgetTest1 createState() {
+    return _WidgetTest1();
+  }
+}
+
+class _WidgetTest1 extends State<WidgetTest1> with AutomaticKeepAliveClientMixin{
+  @override
+  Widget build(BuildContext context) {
+    super.build(context);
+    return Text(widget.text);
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    if (!widget.optionsRef['shouldDispose']) {
+      fail('Widget with text ${widget.text} was disposed when it shouldn\'t.');
```

Comment:
Looking at this it may be better to give the state a variable "disposed", initalize it to false and set it to true in here. In the test you can then expect that the variable is either true or false. That would make this a little more explicit.

Reply:
I wonder how can I save the value of whether the state is disposed. Once it disposed the state is gone, so I can not store that in state. The widget is immutable, I cannot store it in the widget either.

Reply:
done